### PR TITLE
Review

### PR DESCRIPTION
**Pull request by zilton7**
This is our entry for Ruby Project #2: Enumerable Methods

### CODE REVIEW

**Review by smcommits — CHANGES_REQUESTED** [review]
# Status: Changes Required :+1: 
## Hello, @zilton7  and @crappypapa! You guys have done a great job. Your methods almost work like enumerable methods. But, we still need to make some changes to pass all the requirements. 

## Required Changes 
### Note: Maximum cases are failing because you haven't implemented your methods to handle range or regex. Change every method so that they can handle regex and range. It will solve around 90% of the issues. 

### Your methods are missing the following behaviours that are found in original Enumerable methods. 


 - [x] **Enumerables #my_each returns an Enumerator if no block is given**

- [x] **enumerables #my_each when a block is given when self is a range returns the range itself after calling the given block once for each element in self**

- [x] **enumerables #my_each when a block is given when self is a hash behaves like #each calls the given block once for each element in self**

- [x] **enumerables #my_each_with_index returns an enumerator if no block is given**

- [x] **enumerables #my_each_with_index when a block is given when self is a range returns the range itself after calling the given block once for each element in self** 

- [x] **enumerables #my_each_with_index when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self**

- [x] **enumerables #my_select returns an array containing all elements of enum for which the given block returns a true value when a range is given**

- [x] **enumerables #my_select returns an enumerator if no block is given**

- [x] **enumerables #my_all? returns false if the block returns false or nil when a range is given**

- [x] **enumerables #my_all? when no block or argument is given returns true when none of the collection members are false or nil**

- [x] **enumerables #my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern**

- [x] **enumerables #my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern**

- [x] **enumerables #my_any? returns false if the block ever returns a value false or nil when a range is given**

- [x] **enumerables #my_any? when no block or argument is given returns false if at least one of the collection is not true**

- [x] **enumerables #my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern**

- [x] **enumerables #my_none? returns false if the block ever returns true for all elements when a range is given**

- [x] **enumerables #my_none? when a Regex is passed as an argument returns true if none of the collection matches the Regex**

- [x] **enumerables #my_none? when a Regex is passed as an argument returns false if any of the collection matches the Regex**

- [x] **enumerables #my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern**

- [x] **enumerables #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern**

- [x] **enumerables #my_count returns the number of items in enum through enumeration**

- [x] **enumerables #my_count returns the number of items in enum through enumeration when a range is given**

- [x] **enumerables #my_map returns an Enumerator if no block is given**

- [x] **enumerables #my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.**

- [x] **enumerables #my_inject raises a "LocalJumpError" when no block or argument is given**

**If you have any doubt please try to run the original methods for the above mentioned cases and compare the output with your method.**

# 
### Please don't hesitate to contact me on [Slack](https://microverse-students.slack.com/team/U017N7LLZ40) if you need assistance or believe I have overlooked something and made an unfair request.

File: README.md [hunk 1]
```diff
@@ -1 +1,134 @@
...
+- Ruby
+
+## Getting Started
+
+No prerequisites
+
+To get a local copy up and running:
+
+1) Clone the repo or download the Zip folder
+2) Install ruby and run "ruby bubble_sort.rb" in your terminal
```
Comment (cyonii) [inline README.md]: Are you sure this instruction is for this project?

File: enumerable_methods.rb [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    while index < self_arr.length
+      yield(self_arr[index])
+      index += 1
+    end
+    self
+  end
+
+  # p ['a', 'b'].my_each
+  # p (1..5).each.my_each{ |item| p item}
+  # {first: 1, second: 2}.my_each {|key, value| p value}
```
Comment (cyonii) [inline enumerable_methods.rb]: Please do not comment code

File: enumerable_methods.rb [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+      end
+    end
+    return my_select { |item| !yield(item) }.length == to_a.length if block_given?
+
+    my_select(&:!).length == to_a.length if !param && !block_given?
+  end
+
+  # p (1..3).my_none? { |el| el < 1}
+  # p ['a', 'c'].my_none?(/a/)
+  # p ['a', 'c'].my_none?('d')
```
Comment (cyonii) [inline enumerable_methods.rb]: Commented code. Please remove all of the rest in this file

**Review by cyonii — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️✍️

Hello team, you have done a great job so far.
I'm thrilled to see the energy you have put off since the last review.

## To Improve
Kindly review your code and make sure it is passing the following tests. This documentation of Ruby enumerable methods is a good resource to help you finish up this project. [Documentation here](https://ruby-doc.org/core-2.6.5/Enumerable.html)
- [ ] `#my_each` when a block is given when self is a hash behaves like `#each` calls the given block once for each element in self
- [x] `#my_count` counts the number of elements yielding a true value if a block is given
- [x] `#my_inject` raises a `LocalJumpError` when no block or argument is given

Please implement these changes and those requested in the inline comments before resubmitting for review.

Keep your heads up. Happy coding

### You can find me on slack 👉 [CY Kalu](https://microverse-students.slack.com/team/U01494U96FR) if you need further assistance or clarification. I'm always available and will be glad to help.

File: README.md (hunk 1 above)
Comment (zilton7) [reply README.md]: >  #my_each when a block is given when self is a hash behaves like #each calls the given block once for each element in self

Not sure about this, it seems like I get exactly same output compared to original each :)

image

**Review by cyonii — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️✍️


Kindly review the failed test cases, and fix them. You have already come so far, stay motivated.
# 🔋⚡

### Failed examples:

---
#### #my_each
**`#my_each`** > **context:** > when a block is given > **context:** when self is a hash > it behaves like `#each`, calls the given block once for each element in self
```ruby
# Use this to test your method.
# Make a hash, e.g
person = {name: 'John Doe', sex: :male, age: 52}

=begin
Call ruby's #each method on it and immitate the results
for your own methods
=end
```
---
#### #my_inject
```ruby
# These variables may be reused accross the examples
array = [1, 9, 0, 43, 21]
range = Range.new(1, 50)
```

- [x] `#my_inject` raises a `"LocalJumpError"` when no block or argument is given.
**expectation:** `[1, 9, 0, 43, 21].my_inject` should raise `LocalJumpError` as no block is given
- [x] `#my_inject` searches for the longest word in an array of strings
```ruby
words = %w[dog door rod blade]
search = proc { |memo, word| memo.length > word.length ? memo : word }

#Expectation
words.my_inject(&search) == words.inject(&search)
```
- [x] `#my_inject` does not mutate the original array
```ruby
array_clone = array.clone

array.my_inject { |num| num + 1 } # After calling #my_inject

# Expectation
array == array_clone
```
- [x] `#my_inject` when a block is given without an initial value, it combines all elements of enum by applying a binary operation, specified by a block
```ruby
operation = proc { |sum, n| sum + n }

# Expectation
array.my_inject(&operation) == array.inject(&operation)
```
- [x] `#my_inject` when a block is given without an initial value, it combines all elements of enum by applying a binary operation, specified by a block::range
```ruby
# Expectation
operation = { |prod, n| prod * n }
range.my_inject(&operation) == range.inject(&operation)
```
- [x] `#my_inject` when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block
```ruby
actual = range.my_inject(4) { |prod, n| prod * n }
expected = range.inject(4) { |prod, n| prod * n }

#Expectation
actual == expected
```
- [x] `#my_inject` when a symbol is specified without an initial value, it combines each element of the collection by applying the symbol as a named method
```ruby
# Expectations
array.my_inject(:+) == array.inject(:+)
array.my_inject(:*) == array.inject(:*)
```
- [x] `#my_inject` when a symbol is specified without an initial value, it combines each element of the collection by applying the symbol as a named method:range
```ruby
range.my_inject(:*) == range.inject(:*)
```
- [x] `#my_inject` when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method
```ruby
array.my_inject(20, :*) == array.inject(20, :*)
```
- [x] `#my_inject` when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range
```ruby
range.my_inject(2, :*) == range.inject(2, :*)
```

---
#### #multiply_els
- [x] `#multiply_els` accepts an array as an argument and multiplies all the elements of the array together using #my_inject
```ruby
# Expectation
multiply_els [2, 4, 5] == 40
```

>  <br> You `#my_inject` method seemed to be working before now, please fix these issues and resubmit.
Stay positive, and happy coding 💻
<br>

###### You can find me on slack 👉 [CY Kalu](https://microverse-students.slack.com/team/U01494U96FR) if you need further assistance or clarification. I'm always available and will be glad to help.

File: enumerable_methods.rb [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    if block_given? && !proc
+      to_a.my_each { |el| arr << yield(el) }
+    elsif proc
+      to_a.my_each { |el| arr << proc.call(el) }
+    end
+    arr
+  end
+
+  def my_inject(param1 = nil, param2 = nil)
+    raise 'LocalJumpError' unless block_given? || param1 || param2
```
Comment (bini-i) [inline enumerable_methods.rb]: -  make sure my_inject raises LocalJumpError when no block is passed.

**hint**: LocalJumpError is a keyword as it is the name of the exception to be raised.

**Review by bini-i — CHANGES_REQUESTED** [review]
# Status: Changes Required :yellow_circle:    
### Hello @crappypapa and @zilton7 :wave: 

### Good job working on the requirement suggested by previous TSEs. Your project has passed almost all tests except one. Please refer to inline comment below :point_down:.

#### If you have any question kindly reach me on slack @Binyam Hailemeskel

### HAPPY CODING :computer: :blush: 

> #### Reviewed by Binyam Hailemeskel (Part-Time TSE)
> GitHub: [@bini-i](https://github.com/bini-i) 
> Twitter: [@binyamshewa](https://twitter.com/binyamshewa)
> LinkedIn: [LinkedIn](https://www.linkedin.com/in/binyam-hailemeskel-728048151/)

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Olaoluwa Solladoye & Zil Norvilis** 
Good job with the improvements for this project :muscle:  :muscle:
Your methods pass all the tests :clap: :clap: :clap:  
Just a last detail to improve before you can go ahead:

- [x] Fix rubocop offenses (don't edit the rubocop file)
 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

**Review by jubaan — APPROVED** [review]
[![Microverse][microverse-shield]][microverse-url]
<div>
<h1 align="center"> 
  🎉🥳 🎊 🎉 🥳 🥳 🎊 🎉 🥳🥳 🎊 🎉 🥳 🎉 <br />
  🎉🥳 🎊 🎉 🥳 🥳 🎊 🎉 🥳🥳 🎊 🎉 🥳 🎉 <br />
  🥳 🎊 🎉 🥳  APPROVED 🥳 🎊 🎉 🥳 <br />
  🎉🥳 🎊 🎉 🥳 🥳 🎊 🎉 🥳🥳 🎊 🎉 🥳 🎉 <br />
  🎉🥳 🎊 🎉 🥳 🥳 🎊 🎉 🥳🥳 🎊 🎉 🥳 🎉
</h1>
<h2 align="center">
  What’s done is done. What’s gone is gone. <br />
  One of life’s lessons is always moving on. <br />
  It’s okay to look back to see how far you’ve come but <br /> 
  KEEP MOVING FORWARD.
</h2>

<details open="open">
  <summary>Review Summary</summary>
  <ol>
    <li><a href="#optional"> 💡Nice to have and other comments ✍🏼 ✍🏼 </a></li>
    <li><a href="#contact"> 🧔🏼 About your Reviewer </a> (you don't know it... but we are just automated bots 🤖) </li>
    <li><a href="#help"> 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼 </a></li>
  </ol>
</details>
</div>


<h3 id="optional"> 1. 💡 Nice to have and other comments </h3>
<details close="close">
<summary> [click to open] If you have something here, even when approved, it's just a little push to improve your project.</h3></summary>


- [ ] Nothing more to add.


</details>

<h3 id="contact"> 2. 🧔🏼 About your Reviewer </h3>
<details close="close">
<summary> [click to open] Say "hello world!" on Slack: @jubaan</summary>
<br />
<div align="center">
  <a src="https://www.github.com/jubaan">
    Julio Añoveros
  </a>
  <p>
    Julio Añoveros | @jubaan</br></br>
    <b>Full-stack Web Developer</b> and</br>
    🤖 Code Reviewer at <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a>:rocket:
  </p>
</div>
</details>

<h3 id="help"> 3. 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼</h3>
<p> How does it feel? Is nice to advance in the program, right? It keeps you motivated even with all the struggles and takes you one step closer to the goal 🎯!. I wan't you to keep advancing so if you feel stuck and/or need help, remember there is always someone you can reach to keep moving forward, doesn't mater how "dum" you think the questions is.</p>
<details close="close">
  <summary> [click to open] <b>Feel stuck?</b> </summary>
  <br>
  <p> First breath... just do it. Take 15-20 min away from the computer, stretch, meditate, do some exercise, drink water, or do whatever is necessary to distract your mind from the obstacle.</p>
  <p><b>Sometimes you just need to get away to see things with a new perspective.</b></p>
</details>

<details close="close">
  <summary> [click to open] <b>Need help?</b> </summary>
  <br>
  <p> Send me a DM in <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> Students Slack, you'll find me as @jubaan. If I'm not available there are other Code Reviewers you can contact. They'll help as well.</p>
  <p>Remember that <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> has other channels where you can get an answer to your questions:</p>
  <ol>
    <li>The <b>students channel 🧑‍🎓 </b> in slack,</li>
    <li>your <b>teammates 👥 </b>,</li>
    <li>your <b>mentor 👨🏼 👩🏼 </b>, and</li>
    <li><b>Student Success 🏫 </b>.</li>
  </ol>
  <p><b> DON'T hesitate to ask! </b></p>
</details>
<p>If all of the above doesn't help... get a <a href="https://rubberduckdebugging.com/">rubber duck</a> ... it would be great to own one 😄. </p>

<h1 align="center"> See you a the top 🚀!! </h1>

[microverse-shield]: https://img.shields.io/badge/-microverse-%236F23FF?style=for-the-badge
[microverse-url]: https://www.microverse.org/?grsf=kfm2ej
[bootstrap-shield]: https://img.shields.io/badge/Bootstrap-white?style=for-the-badge&logo=bootstrap
[bootstrap-url]: https://www.microverse.org/?grsf=kfm2ej

**Pull request merged**